be1-go/hub/inbox.go

```diff
@@ -0,0 +1,64 @@
+package hub
+
+import (
+	"encoding/base64"
+	"log"
+	"student20_pop/message"
+	"sync"
+	"time"
+)
+
+type inbox struct {
+	mutex sync.RWMutex
+	msgs  map[string]*messageInfo
+}
+
+func createInbox() *inbox {
+	return &inbox{
+		mutex: sync.RWMutex{},
+		msgs:  make(map[string]*messageInfo),
+	}
+}
+
+// addWitnessSig adds a signature of witness to a message of ID `messageID`.
+// if no message has ID `messageID`, does nothing
+func (i *inbox) addWitnessSig(messageID []byte, public message.PublicKey, signature message.Signature) {
+	msg, ok := i.getMessage(messageID)
+	if !ok {
+		// TODO: We received a witness signature before the message itself.
+		// We ignore it for now but it might be worth keeping it until we
+		// actually receive the message
+		log.Printf("failed to find message_id %s for witness message", messageID)
+	}
+	i.mutex.Lock()
+	msg.WitnessSignatures = append(msg.WitnessSignatures, message.PublicKeySignaturePair{
```

Comment:
I don't think this will update the existing message since `getMessage` doesn't return a pointer to the struct

```go
func TestAddWitnessSig_AddWitnessSignature(t *testing.T) {
	inbox := createInbox()

	msg, err := message.NewMessage(message.PublicKey{1, 2, 3}, message.Signature{1, 2, 3}, nil, nil)
	require.NoError(t, err)

	inbox.storeMessage(*msg)

	require.Equal(t, 1, len(inbox.msgs))

	buf, err := base64.URLEncoding.DecodeString("oJYBapM5ZuVrnggAwzQMa3oBLrFSjEQY-hv_JQRgs1U=")
	require.NoError(t, err)

	inbox.addWitnessSig(buf[:], message.PublicKey{4, 5, 6}, message.Signature{7, 8, 9})

	storedMsg, ok := inbox.getMessage(buf[:])
	require.Equal(t, true, ok)

	require.Equal(t, 1, len(storedMsg.WitnessSignatures))
}
```